### PR TITLE
Ledger multiplexer

### PR DESCRIPTION
**Pull request by techlab-lainio**
- blog post of Go wrappers ledger plugin system

### CODE REVIEW

File: content/en/blog/multiplexer/index.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
+---
+date: 2022-09-30
+title: "Ledger Multiplexer"
+linkTitle: "Ledger Multiplexer"
+description: "I implemented a ledger multiplexer in Go for Indy SDK (libindy).
+Be able to replace Indy ledger whit whatever verified data registry. Be able to
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:6]: whit -> with

File: content/en/blog/multiplexer/index.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
+---
+date: 2022-09-30
+title: "Ledger Multiplexer"
+linkTitle: "Ledger Multiplexer"
+description: "I implemented a ledger multiplexer in Go for Indy SDK (libindy).
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:5]: This is a bit hard understand when first there is first-person-form and then it changes like imperative "Be able"

File: content/en/blog/multiplexer/index.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+have automatic ledger backups. Be able to remove the ledger's performance
+bottleneck with just a few simple and standard software engineering practices."
+author: Harri Lainio
+resources:
+- src: "**.{png,jpg}**"
+  title: "Image #:counter"
+---
+
+In this technical blog post, I'll explain how I implemented a plugin system into
+our Indy SDK Go wrapper and then extended it to work as a multiplexer. The plugin
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:16]: Indy SDK Go wrapper -> add link to repository?

File: content/en/blog/multiplexer/index.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+system allows us to use a key/value-based storage system instead of the normal
+Indy ledger. And multiplexing extended the functionality to use multiple data
+sources simultaneously and asynchronously. For instance, we can add a memory cache
+to help Indy ledger, which has been proving to make a considerable difference in
+a multi-tenant agency that can serve thousands of different wallet users simultaneously.
+
+{{< figure src="/blog/2022/09/27/ledger-multiplexer/Plugin.svg" >}}
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:23]: Image is not displaying, change 27 to 30

Comment (lauravuo-techlab) [reply content/en/blog/multiplexer/index.md:23]: There is a typo in the image "FINDY_LEGER" -> d is missing

File: content/en/blog/multiplexer/index.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+- Why was it written in Python when everybody had to know that it would be
+  a performance-critical part of the system?
+- Why did it use ZeroMQ when it used so simple TCP/IP communication? (Was it
+  because of Python?) Anyway, every dependency is a dependency to handle.
+- Why didn't it offer a development run mode from the beginning? For instance, a
+  pseudo node would offer local development and test environment?
+- Why it didn't offer a straightforward own separated API? With Indy SDK, you had to build
+  each transaction with three separate functions that weren't general but
+  entity-specific, like `indy_build_cred_def_request()`.
+- Why the transaction model was so unclear and 'hid' from the rest of the
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:44]: Question mark missing.

File: content/en/blog/multiplexer/index.md [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+plugin routes transactions to the authentic Indy ledger. We will show in this
+post how it's implemented and used.
+
+## Reduce Setup Complexity
+
+Indy ledger has been bugging me since we met each other:
+
+- Why was it written in Python when everybody had to know that it would be
+  a performance-critical part of the system?
+- Why did it use ZeroMQ when it used so simple TCP/IP communication? (Was it
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:37]: Suggestion: Why *it was implemented* with ZeroMQ when it used...

File: content/en/blog/multiplexer/index.md [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+Both functions check if they can find information from the wallet, and the
+ledger is the backup. *For those listening to our presentation on Hyperledger
+Global Forum, I mistakenly said that I used -1 for the wallet handle, which is
+incorrect. Sorry about that. (1-0, once again, for documentation.)*
+
+I discovered that I could enumerate our ledger connection handles
+starting from -1 and going down like -1, -2, and so forth. So I didn't need any extra
+maps to convert connection handles, which would add complexity and affect
+performance. I could give connection handles with negative values to
+the above functions, and `libindy` was OK.
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:108]: Suggestion: `libindy` was OK -> accepted that

File: content/en/blog/multiplexer/index.md [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+- `pairwise`
+- `crypto`
+- `ledger`
+- `pool`, this was very important, because it gave as the entry point to
+above layer. The layer was using our Go wrapper.
+
+The following picture illustrates the whole system where the ledger connection *pool* is replaced
+with our own `pool` package.
+
+{{< figure src="/blog/2022/09/27/ledger-multiplexer/Main.svg" >}}
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:305]: Image is not displaying, change 27 to 30

File: content/en/blog/multiplexer/index.md [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+
+sid, scJSON = try.To2(ledger.ReadSchema(pool, stewardDID, sid))
+```
+
+All of this just by following sound software engineering practices like:
+- abstraction hiding
+- polymorphism
+- modular structures
+
+I hope this was a help. Until the next time, see you!
```
Comment (lauravuo-techlab) [inline content/en/blog/multiplexer/index.md:326]: Suggestion: I hope this was helpful.

**Review by lauravuo-techlab — APPROVED** [review]
Good text, I think there is much more background and clarifications than was even needed 👍 

Added a few comments about the missing images & other minor findings.

File: content/en/blog/multiplexer/index.md (hunk 2 above)
Comment (techlab-lainio) [reply content/en/blog/multiplexer/index.md:5]: agree, but trusted Grammarly :-D Have to read I the 100th time and just make a decision.

File: content/en/blog/multiplexer/index.md (hunk 4 above)
Comment (techlab-lainio) [reply content/en/blog/multiplexer/index.md:23]: Cool, so cool! thank you so much!

**Comment by techlab-lainio** [comment]
Agree about the 'amount of' -background comment. My first thought was that I remove some parts of it but then I thought that at least this can be documentation for whoever has to maintain or study our plugin system in the Go wrapper.

**Pull request merged**